ttnn/cpp/ttnn/operations/ccl/erisc_datamover_builder.cpp

```diff
@@ -0,0 +1,414 @@
+
+    return FabricEriscDatamoverBuilder(
+        ethernet_core,
+        device->ethernet_core_from_logical_core(ethernet_core).x,
+        device->ethernet_core_from_logical_core(ethernet_core).y,
+        local_chip_id,
+        peer_chip_id,
+
+        receiver_channel_downstream_flow_control_semaphore_id,
+        sender_channel_0_flow_control_semaphore_id,
+        sender_channel_1_flow_control_semaphore_id,
+        sender_channel_0_connection_semaphore_id,
+        sender_channel_1_connection_semaphore_id,
+        sender_channel_0_buffer_index_semaphore_id,
+        sender_channel_1_buffer_index_semaphore_id,
+
+        config);
+}
+
+SenderWorkerAdapterSpec FabricEriscDatamoverBuilder::build_connection_to_worker_channel() const {
+    return SenderWorkerAdapterSpec {
+        this->my_noc_x,
+        this->my_noc_y,
+        this->local_sender_channel_0_buffer_address,
+        this->sender_0_num_buffers,
+        this->sender_channel_0_flow_control_semaphore_id,
+        this->sender_channel_0_connection_semaphore_id,
+        FabricEriscDatamoverConfig::sender_channel_0_worker_connection_info_address,
+        this->config.channel_buffer_size_bytes,
+        this->sender_channel_0_buffer_index_semaphore_id
+    };
+}
+
+
+SenderWorkerAdapterSpec FabricEriscDatamoverBuilder::build_connection_to_fabric_channel() const {
+    return SenderWorkerAdapterSpec {
+        this->my_noc_x,
+        this->my_noc_y,
+        this->local_sender_channel_1_buffer_address,
+        this->sender_1_num_buffers,
+        this->sender_channel_1_flow_control_semaphore_id,
+        this->sender_channel_1_connection_semaphore_id,
+        FabricEriscDatamoverConfig::sender_channel_1_worker_connection_info_address,
+        this->config.channel_buffer_size_bytes,
+        this->sender_channel_1_buffer_index_semaphore_id
+    };
+}
+
+void FabricEriscDatamoverBuilder::connect_to_downstream_edm(FabricEriscDatamoverBuilder const& downstream_edm) {
+    auto const& adapter_spec = downstream_edm.build_connection_to_fabric_channel();
+
+    log_trace(tt::LogTest, "Connecting to downstream EDM at x={}, y={}", adapter_spec.edm_noc_x, adapter_spec.edm_noc_y);
+
+    this->downstream_edm_noc_x = adapter_spec.edm_noc_x;
+    this->downstream_edm_noc_y = adapter_spec.edm_noc_y;
+    this->downstream_edm_buffer_base_address = adapter_spec.edm_buffer_base_addr;
+    this->downstream_edm_semaphore_address = adapter_spec.edm_l1_sem_addr;
+    this->downstream_edm_worker_registration_address = adapter_spec.edm_connection_handshake_addr;
+    this->downstream_edm_worker_location_info_address = adapter_spec.edm_worker_location_info_addr;
+    this->downstream_sender_channel_buffer_index_semaphore_id = adapter_spec.buffer_index_semaphore_id;
+}
+
+
+
+EdmLineFabricOpInterface::EdmLineFabricOpInterface (std::vector<Device*> const& device_sequence, std::vector<Program*> const& program_sequence, std::optional<size_t> desired_num_links) :
```

Comment:
This would copy an incoming vector of Device* and Program*, which does not look like it was intended.

Reply:
I intentionally chose to copy in the object here because some of the details about the lifetimes of the passed in vectors are to be worked out so I just wanted to avoid any headaches/bugs in case I forgot that the vectors weren't copied (in this hypothetical alternate reality).